### DISCUSSION TITLE
Variant directive on existing variant-based components

### DISCUSSION THREAD

**Discussion by owensgit** [Legal-and-General/canopy#292] [Ideas] [answered]
I'm currently building a variant directive, which works like this:

```html
<div lgVariant="success">
  I have the success variant styling, yay!
</div>
```

_Work in progress_ screenshot 😉 
![Screenshot 2021-04-01 at 16 56 23](https://user-images.githubusercontent.com/1943532/113321196-42d34f00-930b-11eb-9924-38d0f60a4c23.png)

It's working great, but I have an issue with our existing components, like the Alert, Details and Form Validation, which currently roll their own variant inputs and apply their own classes.

To avoid a breaking change, I'd like `lgVariant` to be used under the hood, when the `variant` Input on each of these components is used. E.g. `<lg-alert variant="success">`.

The problem is that Angular currently does not support using directives on host elements, which is where we'd need to apply this styling (they do intend to add [this feature in on their future roadmap](https://angular.io/guide/roadmap#support-adding-directives-to-host-elements) though).

### Workaround 1

One possible workaround would be to add an inner container, where the directive could be applied. E.g. for the alert component HTML:

```html
<div class="lg-alert__inner" [lgVariant]="variant">
```

This isn't nice of course, as it goes against our pattern of styling the host to avoid superfluous DOM elements, and it would break the SCSS only/flat HTML implementation.

### Workaround 2
Another possible workaround could be to expose methods on the directive class which the component could use. Something along the lines of...

```js
  // Alert component
  _variant: AlertVariant;
  @Input()
  set variant(variant: AlertVariant) {
    if (this._variant) {
      this.variantDirective.applyVariant(variant, this.hostElement.nativeElement)
    }
  }
```

That doesn't feel quite right either, and requires adding a special method just for this usecase - but potentially less messy than the inner container.

### Workaround 3

Alternatively, we could use the new global variant CSS classes introduced alongside the directive, such as `lg-variant--success` and `lg-variant--info` instead of the specific ones that each component has, and essentially just keep the current variant logic in each component - i.e. how it adds/removes variant classes when the Input is set. E.g. 

```js
// Alert component
_variant: Variant;
@Input()
set variant(variant: Variant) {
  if (this._variant) {
    this.renderer.removeClass(
      this.hostElement.nativeElement,
      `lg-variant--${this._variant}`, // uses global class
    );
  }
  this.renderer.addClass(this.hostElement.nativeElement, `lg-variant--${variant}`); // uses global class
  this._variant = variant;
}
```

My initial thoughts are that **Workaround 3** is probably the way to go, as it just means some duplication of logic and doesn't add anything messy. We could then mark the variant inputs on each of these components for deprecation, and then at some point require adopters to use the directive.

Thanks for reading! Let me know what you think - or if you have another idea 🙂

**Comment by andy-polhill** [comment]
I was wondering whether the solution in here is another way of doing what we need?
https://stackoverflow.com/questions/41298168/how-to-dynamically-add-a-directive

It not option 3 get's us to where we need to be short term. In the longer term we need to decide if it's ok to have both the variant input and the utility directive. Or if we eventually want to remove the variant input with a breaking change.

**Reply by owensgit** [reply]
I'm apprehensive about that Stackoverfow solution as it circumvents Angular's dependency injection - and doesn't right to me - but would be willing to discuss. I'll reach out and see if anyone else has more views on that.

**Reply by owensgit** [reply]
I reached out and confirmed that the Stackoverfow solution is not a good idea, as circumventing Angular's dependency injection can lead to some unwanted side effects. There are also limitations in that approach as it doesn't allow you to easily remove the directive once applied.

**Comment by owensgit** [comment, marked as answer]
The chosen approach is **Workaround 3**

> Alternatively, we could use the new global variant CSS classes introduced alongside the directive, such as lg-variant--success and lg-variant--info instead of the specific ones that each component has, and essentially just keep the current variant logic in each component.

Implemented in this PR: https://github.com/Legal-and-General/canopy/pull/298
